File: sdk/python/src/axern_sdk/node/computer_use_protocol.py
"""Computer-use protobuf request builders shared by sync and async clients."""

from __future__ import annotations

from axern.node.sandbox.v1 import node_pb2
from axern_sdk.node.models import (
    ComputerUseDependencyStatus,
    ComputerUseDisplay,
    ComputerUseRegion,
    ComputerUseScreenshot,
    ComputerUseStatus,
)


def computer_use_status(response: node_pb2.ComputerUseStatusResponse) -> ComputerUseStatus:
    return ComputerUseStatus(
        available=bool(response.available),
        display=response.display,
        backend=response.backend,
        reason=response.reason,
        dependencies=tuple(
            ComputerUseDependencyStatus(
                name=item.name,
                available=bool(item.available),
                reason=item.reason,
            )
            for item in response.dependencies
        ),
    )


def computer_use_display(response: node_pb2.ComputerUseDisplayResponse) -> ComputerUseDisplay:
    return ComputerUseDisplay(
        display=response.display,
        backend=response.backend,
        width=response.width,
        height=response.height,
    )


def computer_use_screenshot(response: node_pb2.ComputerUseScreenshotResponse) -> ComputerUseScreenshot:
    return ComputerUseScreenshot(data=bytes(response.data), content_type=response.content_type)


def computer_use_screenshot_request(
    allocation_id: str,
    *,
    show_cursor: bool,
    region: ComputerUseRegion | None,
    format: str,
    quality: int,
    scale: float,
) -> node_pb2.ComputerUseScreenshotRequest:
    request = node_pb2.ComputerUseScreenshotRequest(
        allocation_id=allocation_id,
        show_cursor=show_cursor,
        format=format,
        quality=quality,
        scale=scale,
    )
    if region is not None:
        request.region.CopyFrom(computer_use_region(region))
    return request


def computer_use_region(region: ComputerUseRegion) -> node_pb2.ComputerUseRegion:
    return node_pb2.ComputerUseRegion(
        x=region.x,
        y=region.y,
        width=region.width,
        height=region.height,
    )

File: runtime/axnoded/internal/runtime/sandboxd/computer_use_protocol.go
package sandboxd

type ComputerUseStatusResponse struct {
	Available    bool                          `json:"available"`
	Display      string                        `json:"display,omitempty"`
	Backend      string                        `json:"backend,omitempty"`
	Reason       string                        `json:"reason,omitempty"`
	Dependencies []ComputerUseDependencyStatus `json:"dependencies,omitempty"`
}

type ComputerUseDependencyStatus struct {
	Name      string `json:"name"`
	Available bool   `json:"available"`
	Reason    string `json:"reason,omitempty"`
}

type ComputerUseRegion struct {
	X      int `json:"x"`
	Y      int `json:"y"`
	Width  int `json:"width"`
	Height int `json:"height"`
}

type ComputerUseScreenshotRequest struct {
	ShowCursor bool               `json:"showCursor,omitempty"`
	Region     *ComputerUseRegion `json:"region,omitempty"`
	Format     string             `json:"format,omitempty"`
	Quality    int                `json:"quality,omitempty"`
	Scale      float64            `json:"scale,omitempty"`
}

type ComputerUseScreenshotResponse struct {
	Data        []byte
	ContentType string
}

type ComputerUseDisplayResponse struct {
	Display string `json:"display,omitempty"`
	Backend string `json:"backend,omitempty"`
	Width   int    `json:"width,omitempty"`
	Height  int    `json:"height,omitempty"`
}

type ComputerUseMouseRequest struct {
	Action    string `json:"action,omitempty"`
	X         int    `json:"x,omitempty"`
	Y         int    `json:"y,omitempty"`
	ToX       int    `json:"toX,omitempty"`
	ToY       int    `json:"toY,omitempty"`
	Button    string `json:"button,omitempty"`
	Direction string `json:"direction,omitempty"`
	Amount    int    `json:"amount,omitempty"`
}

type ComputerUseKeyboardRequest struct {
	Text    string   `json:"text,omitempty"`
	Key     string   `json:"key,omitempty"`
	Keys    []string `json:"keys,omitempty"`
	DelayMS int      `json:"delayMs,omitempty"`
}
